### ISSUE TITLE
Non-executable binary files have executable bit set when console quits

### ISSUE THREAD

**Issue by schuhschuh** [openmole/openmole#42]
The CopyFileHook seems to set the executable bit of the copied non-text files. Even though these are binary files, they are not executable and therefore this bit should not be set.

**Comment by schuhschuh** [comment]
While the initial output files have no correct mode (i.e., no executable bit). I noticed that the bit was suddenly set at a later time point again which doesn't seem to be related to the CopyFileHook, but with the use of the file as input of another task ?!?

Here's my terminal output while I was monitoring the creation of output files. You can notice that the files in the "affine" directory initially had no executable bit, but some time later it was set.
```
as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/affine/
total 11K
drwxr-xr-x 2 as12312 vip   5 May  6 05:07 .
drwxr-xr-x 4 as12312 vip   4 May  6 05:05 ..
-rw-r--r-- 1 as12312 vip 131 May  6 05:07 mni305,1000.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:05 mni305,1001.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:07 mni305,1002.dof.gz

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/
total 12K
drwxr-xr-x 4 as12312 vip 4 May  6 05:05 .
drwxr-xr-x 4 as12312 vip 4 May  6 05:03 ..
drwxr-xr-x 2 as12312 vip 5 May  6 05:07 affine
drwxr-xr-x 2 as12312 vip 5 May  6 05:03 rigid

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/
total 13K
drwxr-xr-x 5 as12312 vip 5 May  6 05:08 .
drwxr-xr-x 4 as12312 vip 4 May  6 05:03 ..
drwxr-xr-x 2 as12312 vip 5 May  6 05:07 affine
drwxr-xr-x 2 as12312 vip 5 May  6 05:08 initial
drwxr-xr-x 2 as12312 vip 5 May  6 05:03 rigid

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/initial/
total 11K
drwxr-xr-x 2 as12312 vip   5 May  6 05:08 .
drwxr-xr-x 5 as12312 vip   5 May  6 05:08 ..
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1000,1001.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1000,1002.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1001,1002.dof.gz

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/initial/
total 11K
drwxr-xr-x 2 as12312 vip   5 May  6 05:08 .
drwxr-xr-x 5 as12312 vip   5 May  6 05:08 ..
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1000,1001.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1000,1002.dof.gz
-rw-r--r-- 1 as12312 vip 131 May  6 05:08 1001,1002.dof.gz

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/
total 15K
drwxr-xr-x 5 as12312 vip 5 May  6 05:08 .
drwxr-xr-x 4 as12312 vip 4 May  6 05:03 ..
drwxr-xr-x 2 as12312 vip 5 May  6 05:07 affine
drwxr-xr-x 2 as12312 vip 5 May  6 05:08 initial
drwxr-xr-x 2 as12312 vip 5 May  6 05:03 rigid

as12312@merapi:/homes/as12312/Data/Registrations
> l Workspace/MAC2012/dofs/affine/
total 11K
drwxr-xr-x 2 as12312 vip   5 May  6 05:07 .
drwxr-xr-x 5 as12312 vip   5 May  6 05:08 ..
-rwxr--r-- 1 as12312 vip 131 May  6 05:07 mni305,1000.dof.gz
-rwxr--r-- 1 as12312 vip 131 May  6 05:05 mni305,1001.dof.gz
-rwxr--r-- 1 as12312 vip 131 May  6 05:07 mni305,1002.dof.gz
```

**Comment by schuhschuh** [comment]
I would consider this issue even a bit more sever after noticing that OpenMOLE not only sets the execution bit of the output files at some point again, but even modifies the permissions of the input files which were initially sampled by a ```SelectFileDomain``` in the exploration task! In particular, I noticed that input files had the executable bit set and removed it before starting the workflow. Once OpenMOLE finished, the executable bit of these input files was set again.

**Comment by romainreuillon** [comment]
Could you try to print the stack and some info in the mode_= method of the org.openmole.tool.file package object to try to understand when this flag is set?

**Comment by schuhschuh** [comment]
To print the stack trace, do I have to throw an exception within this method and catch it right away?

```scala
try throw new Exception("bla")
catch {
  case e: Exception => e.printStackTrace()
}
```

**Comment by schuhschuh** [comment]
The ```FileDecorator.mode_(m: Int)``` function sets the executable bit of an input file right after the task output was printed. Here's the stack trace:
```
Setting executable bit of file /Users/as12312/Data/Registrations/Template/mni305.cor.nii.gz
java.lang.Exception
        at org.openmole.tool.file.package$FileDecorator.mode_$eq(package.scala:233)
        at org.openmole.tool.file.package$.org$openmole$tool$file$package$$authorizeLS$1(package.scala:417)
        at org.openmole.tool.file.package$$anonfun$org$openmole$tool$file$package$$recurse$2.apply(package.scala:421)
        at org.openmole.tool.file.package$$anonfun$org$openmole$tool$file$package$$recurse$2.apply(package.scala:421)
        at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
        at org.openmole.tool.file.package$.org$openmole$tool$file$package$$recurse(package.scala:421)
        at org.openmole.tool.file.package$FileDecorator.applyRecursive(package.scala:396)
        at org.openmole.plugin.task.external.ExternalTask$class.fetchOutputFiles(ExternalTask.scala:132)
        at org.openmole.plugin.task.scala.ScalaTask.fetchOutputFiles(ScalaTask.scala:35)
        at org.openmole.plugin.task.jvm.JVMLanguageTask$class.process(JVMLanguageTask.scala:45)
        at org.openmole.plugin.task.scala.ScalaTask.process(ScalaTask.scala:35)
        at org.openmole.core.workflow.task.Task$$anonfun$perform$1.apply(Task.scala:46)
        at org.openmole.core.workflow.task.Task$$anonfun$perform$1.apply(Task.scala:46)
        at org.openmole.core.workflow.tools.InputOutputCheck$class.perform(InputOutputCheck.scala:91)
        at org.openmole.plugin.task.scala.ScalaTask.perform(ScalaTask.scala:35)
        at org.openmole.core.workflow.task.Task$class.perform(Task.scala:46)
        at org.openmole.plugin.task.scala.ScalaTask.perform(ScalaTask.scala:35)
        at org.openmole.core.workflow.job.MoleJob.perform(MoleJob.scala:99)
        at org.openmole.core.workflow.execution.local.LocalExecuter$$anonfun$run$1.apply(LocalExecuter.scala:66)
        at org.openmole.core.workflow.execution.local.LocalExecuter$$anonfun$run$1.apply(LocalExecuter.scala:58)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.openmole.core.workflow.execution.local.LocalExecuter.run(LocalExecuter.scala:58)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by schuhschuh** [comment]
P.S.: Line numbers in ```file/package.scala``` are off by +14 below the definition of the ```mode_``` functions.

**Comment by schuhschuh** [comment]
The bug is in FileDecorator.recurse.authorizeLS:

```scala
val originalMode = f.mode
```

should be ```f.realFile.mode```.

**Comment by romainreuillon** [comment]
BTW, the authorizeLS(f) is not at right place it should something like that:
```scala   
if (file.isDirectory && !file.isSymbolicLink) authorizeLS(file) {
      for (f ← Option(file.listFiles).getOrElse(Array.empty))  { recurse(f)(operation, stopPath) }
}
```

**Comment by schuhschuh** [comment]
Ah, now I understand what it is even good for.

**Comment by schuhschuh** [comment]
Also, my first conclusion that the bug would be the ```f.mode``` is wrong as well... because this function is defined in ```FileDecorator.mode``` as follows:
```scala
def mode =
      { if (Files.isReadable(file.realFile)) READ_MODE else 0 } |
      { if (Files.isWritable(file.realFile)) WRITE_MODE else 0 } |
      { if (Files.isExecutable(file.realFile)) EXEC_MODE else 0 }
```

**Comment by schuhschuh** [comment]
Your change was the actual fix.

Besides that, I've modified the mode setter to operate on the linked file rather the link itself (which however is anyway the case I believe) as proper counterpart to the mode getter. These changes just make it more obvious and save a few unnecessary "toFile" "toPath" conversions.

**Comment by schuhschuh** [comment]
When I run the whole workflow 'till the end on my Mac using a LocalEnvironment, all was fine. But now that I run the latest master HEAD revision on my workstation with Ubuntu 12.04 and CondorEnvironment, I am seeing this issue again.

(Slapping myself)

Of course! Testing it in a LocalEnvironment didn't make sense because it doesn't use any symbolic links.

Re-opening the issue and trying to figure the true cause of the problem...

**Comment by schuhschuh** [comment]
Here's another stack trace of when an input file's mode is changed. Couldn't find an output for output files which are input to other tasks that have the bit set once they are consumed (not produced), but maybe their mode is based on this input file?
```
Setting the executable bit of /vol/medic01/users/as12312/Data/Registrations/Template/mni305.cor.nii.gz
java.lang.Exception
        at org.openmole.tool.file.package$FileDecorator.mode_$eq(package.scala:237)
        at org.openmole.core.batch.message.package$ReplicatedFileDecorator.download(package.scala:60)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$getContextResults$1$$anonfun$1.apply(GetResultActor.scala:131)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$getContextResults$1$$anonfun$1.apply(GetResultActor.scala:130)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$getContextResults$1.apply(GetResultActor.scala:129)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$getContextResults$1.apply(GetResultActor.scala:127)
        at org.openmole.core.workspace.Workspace.withTmpFile(Workspace.scala:217)
        at org.openmole.core.batch.refresh.GetResultActor.getContextResults(GetResultActor.scala:127)
        at org.openmole.core.batch.refresh.GetResultActor.getResult(GetResultActor.scala:71)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$receive$1$$anonfun$apply$mcV$sp$1.apply(GetResultActor.scala:50)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$receive$1$$anonfun$apply$mcV$sp$1.apply(GetResultActor.scala:48)
        at org.openmole.core.batch.control.UsageControl$class.tryWithToken(UsageControl.scala:28)
        at org.openmole.plugin.environment.ssh.SSHPersistentStorage$$anon$2.tryWithToken(SSHPersistentStorage.scala:47)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$receive$1.apply$mcV$sp(GetResultActor.scala:48)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$receive$1.apply(GetResultActor.scala:46)
        at org.openmole.core.batch.refresh.GetResultActor$$anonfun$receive$1.apply(GetResultActor.scala:46)
        at org.openmole.core.batch.refresh.package$.withRunFinalization(package.scala:23)
        at org.openmole.core.batch.refresh.GetResultActor.receive(GetResultActor.scala:46)
        at org.openmole.core.batch.refresh.JobManager$DispatcherActor$.receive(JobManager.scala:84)
        at org.openmole.core.batch.refresh.JobManager$$anonfun$1$$anon$1.run(JobManager.scala:63)
```
Besides this occasion, there are plenty of the following for all the .jar files linked in the "persistent" directory:
```
Setting the executable bit of /vol/medic01/users/as12312/Code/OpenMOLE/openmole/bin/openmole/target/assemble/openmole-plugins/org-openmole-plugin-grouping-batch_2.11-5.0-SNAPSHOT.jar
java.lang.Exception
        at org.openmole.tool.file.package$FileDecorator.mode_$eq(package.scala:237)
        at org.openmole.core.batch.message.package$ReplicatedFileDecorator.download(package.scala:60)
        at org.openmole.runtime.runtime.Runtime.org$openmole$runtime$runtime$Runtime$$getReplicatedFile$1(Runtime.scala:85)
        at org.openmole.runtime.runtime.Runtime$$anonfun$5.apply(Runtime.scala:99)
        at org.openmole.runtime.runtime.Runtime$$anonfun$5.apply(Runtime.scala:98)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:322)
        at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:978)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
        at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:47)
        at scala.collection.SetLike$class.map(SetLike.scala:92)
        at scala.collection.AbstractSet.map(Set.scala:47)
        at org.openmole.runtime.runtime.Runtime.apply(Runtime.scala:98)
        at org.openmole.runtime.runtime.SimExplorer$$anonfun$start$1.apply(SimExplorer.scala:87)
        at org.openmole.runtime.runtime.SimExplorer$$anonfun$start$1.apply(SimExplorer.scala:77)
        at scala.Option.foreach(Option.scala:257)
        at org.openmole.runtime.runtime.SimExplorer.start(SimExplorer.scala:77)
        at org.openmole.runtime.runtime.SimExplorer.start(SimExplorer.scala:37)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:110)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:79)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:353)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:180)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:629)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:584)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1438)
        at org.eclipse.equinox.launcher.Main.main(Main.java:1414)
```

**Comment by schuhschuh** [comment]
But there must be further occasions of file mode changes not covered by my debug output because further input files have the executable bit set but are never mentioned in the output.

**Comment by schuhschuh** [comment]
Strange... after I added the debug output to the download function, the mode of all input files remains correct. Only some output files are affected (those which depend on the mni305.cor.nii.gz input file whose executable bit is reported above as being set at one time point and not set at another...).

Or not.... the mode of the input image files (e.g., 1000_3.nii.gz) **is** still modified again (executable bit set) once the entire workflow completed and after I quit the OpenMOLE console. I think not before that.

```
May 09, 2015 5:14:16 AM org.openmole.core.batch.environment.BatchJobWatcher update
FINE: Watch jobs 0
May 09, 2015 5:15:16 AM org.openmole.core.batch.environment.BatchJobWatcher update
FINE: Watch jobs 0
May 09, 2015 5:16:16 AM org.openmole.core.batch.environment.BatchJobWatcher update
FINE: Watch jobs 0

OpenMOLE>
```

**Comment by JusteRaimbault** [comment]
Inactive for too long, closing this
